File: scripts/crosscheck_scan.js
#!/usr/bin/env node
/* Check app.html's spike scan against the Python, at every definition in a reference
 * file written by crosscheck_scan.py.
 *
 *     python3 scripts/crosscheck_scan.py /tmp/scan_ref.json
 *     node scripts/crosscheck_scan.js /tmp/scan_ref.json
 *
 * The functions are not copied out of app.html — they are *run* out of it. The whole
 * <script> is evaluated inside a small sandbox and the pieces that matter are handed
 * back, so this cannot pass while testing a stale copy of code that has since changed,
 * which is the failure mode a copied-out version would have.
 *
 * The page's boot sequence runs too, and is expected to fail: its first act is to fetch
 * data/days.json, and the stub below rejects. That happens after an await, so every
 * top-level declaration is already in place by the time the rejection lands.
 */
const fs = require('fs')
const path = require('path')

const ROOT = path.dirname(__dirname)
const refPath = process.argv[2]
if (!refPath) {
  console.error('usage: node scripts/crosscheck_scan.js <reference.json>')
  process.exit(2)
}

process.on('unhandledRejection', () => {})   // the boot IIFE, see above

const html = fs.readFileSync(path.join(ROOT, 'app.html'), 'utf8')
const m = html.match(/<script>([\s\S]*)<\/script>/)
if (!m) { console.error('no <script> in app.html'); process.exit(2) }

const noop = () => {}
const el = { textContent: '', innerHTML: '', hidden: false, classList: { toggle: noop, add: noop, remove: noop },
             setAttribute: noop, getAttribute: () => null, querySelector: () => null,
             querySelectorAll: () => [], addEventListener: noop, remove: noop, focus: noop }
const sandbox = {
  document: { getElementById: () => el, querySelector: () => null, querySelectorAll: () => [],
              addEventListener: noop, createElement: () => el, body: el, documentElement: el },
  window: { addEventListener: noop, matchMedia: () => ({ matches: false, addEventListener: noop }) },
  location: { search: '', pathname: '/', href: '/' },
  history: { pushState: noop, replaceState: noop },
  localStorage: { getItem: () => null, setItem: noop },
  fetch: () => Promise.reject(new Error('offline by design')),
  requestAnimationFrame: noop,
}

/* The trailing expression hands back the entities under test plus one setter, which
 * closes over the module's own `let` bindings — that is how the archive gets in without
 * the functions taking it as an argument and drifting from the page's version. */
const factory = new Function(...Object.keys(sandbox), `
${m[1]}
;return {
  scanEvents, scanFinding, sha256hex, pyRound, SPK0,
  load(days, snaps, prices){
    DAYS = days; PRICES = prices;
    for (const k in MAPS) delete MAPS[k];
    for (const k in snaps) SNAP[k] = snaps[k];
  },
};`)
const app = factory(...Object.values(sandbox))

if (sha256check(app) !== true) process.exit(1)

const days = JSON.parse(fs.readFileSync(path.join(ROOT, 'data/days.json'), 'utf8'))
const prices = JSON.parse(fs.readFileSync(path.join(ROOT, 'data/prices.json'), 'utf8'))
const snaps = {}
for (const d of days) {
  const p = path.join(ROOT, 'data/apewisdom', d + '.json')
  if (fs.existsSync(p)) snaps[d] = JSON.parse(fs.readFileSync(p, 'utf8'))
}
app.load(days, snaps, prices)

const ref = JSON.parse(fs.readFileSync(refPath, 'utf8'))
const FIELDS = ['n', 'median', 'trimmed_mean', 'mean', 'win_rate']
let checked = 0
const bad = []

for (const key of Object.keys(ref)) {
  const [floor, mult, look, gap] = key.split('-').map(Number)
  const events = app.scanEvents({ floor, mult, look, gap })
  const h = app.scanFinding(events, [20]).horizons[0] || {}
  const R = ref[key]
  const diffs = []

  const mine = events.map(e => e.tk + '|' + e.d).sort()
  if (mine.length !== R.keys.length || mine.some((x, i) => x !== R.keys[i])) {
    const a = new Set(mine), b = new Set(R.keys)
    diffs.push(`event set: ${mine.length} vs ${R.keys.length}` +
      ` · only in js ${[...a].filter(x => !b.has(x)).slice(0, 3)}` +
      ` · only in py ${[...b].filter(x => !a.has(x)).slice(0, 3)}`)
  }
  if (R.gap_median !== h.gap_median) diffs.push(`gap_median: py ${R.gap_median} js ${h.gap_median}`)
  for (const side of ['spike', 'placebo']) {
    for (const f of FIELDS) {
      const p = R[side] && R[side][f], j = h[side] && h[side][f]
      if (p !== j) diffs.push(`${side}.${f}: py ${p} js ${j}`)
    }
  }
  checked++
  if (diffs.length) bad.push({ key, diffs })
}

function sha256check(a) {
  const want = 'ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad'
  const got = a.sha256hex('abc')
  if (got !== want) { console.error(`sha256hex('abc') = ${got}\n  expected ${want}`); return false }
  /* Multi-block and non-ASCII, because the padding maths is where a hand-written digest
   * goes wrong, and every placebo pick runs through it. */
  const crypto = require('crypto')
  for (const s of ['', 'a'.repeat(55), 'a'.repeat(56), 'a'.repeat(64), 'a'.repeat(200),
                   'BRK.B|2026-08-09|20', '证券|2021-05-10|5']) {
    const ours = a.sha256hex(s), theirs = crypto.createHash('sha256').update(s, 'utf8').digest('hex')
    if (ours !== theirs) { console.error(`sha256hex(${JSON.stringify(s)}) diverges:\n  ${ours}\n  ${theirs}`); return false }
  }
  if (a.pyRound(12.5) !== 12 || a.pyRound(13.5) !== 14 || a.pyRound(-12.5) !== -12) {
    console.error('pyRound does not break ties to even like Python round()'); return false
  }
  return true
}

if (bad.length) {
  console.error(`${bad.length} of ${checked} definitions disagree with the Python:\n`)
  for (const b of bad.slice(0, 10)) console.error(`  ${b.key}\n    ${b.diffs.join('\n    ')}`)
  process.exit(1)
}
console.log(`app.html reproduces the Python at all ${checked} definitions ` +
            `(event sets and every describe() field, both sides)`)

File: scripts/crosscheck_scan.py
#!/usr/bin/env python3
"""Reference results for the browser's spike scan, at many definitions -> a JSON file.

app.html reimplements precompute_events.py and finding.py in JavaScript so the spike
definition can be moved with sliders. selfCheck() in the browser proves the port at the
*default* definition, against data/events.json and data/finding.json — but the default
is exactly the one setting where a golden file exists. Everywhere else, the port could
drift and nothing would say so, which matters because the whole point of the control is
that people will leave the default.

So: this writes what the Python produces at a spread of settings, and crosscheck_scan.js
checks the JavaScript against it. That pair found a real divergence the golden files
could not — Python's round() breaks ties to even and Math.round breaks them upward, so a
win rate of 2 out of 16 came out 12 here and 13 there. Nothing at the default reaches an
exact tie.

The settings cover the slider's own range and its eight corners, plus a seeded random
spread, so a rerun compares the same points.

    python3 scripts/crosscheck_scan.py /tmp/scan_ref.json          # 40 settings
    python3 scripts/crosscheck_scan.py /tmp/scan_ref.json --n 120  # slower, wider
"""
import json, os, random, statistics as st, sys

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
sys.path.insert(0, os.path.join(ROOT, "scripts"))
import finding as F                                    # noqa: E402  reuse, never re-derive

# The slider's own stops, so nothing here is unreachable through the UI.
FLOORS = range(5, 205, 5)
MULTS = [1.5 + 0.5 * i for i in range(14)]
LOOKS = (5, 40)
GAPS = (0, 20)
CORNERS = [(30, 3, 20, 5),                             # the shipped definition
           (5, 1.5, 5, 0), (200, 8, 40, 20),           # loosest / tightest
           (5, 8, 40, 0), (200, 1.5, 5, 20),
           (5, 1.5, 40, 20), (200, 8, 5, 0),
           (185, 3.5, 40, 5)]                          # the banker's-rounding case


def load_archive():
    d = os.path.join(ROOT, "data", "apewisdom")
    dates, maps = [], {}
    for name in sorted(os.listdir(d)):
        if not name.endswith(".json"):
            continue
        with open(os.path.join(d, name)) as f:
            snap = json.load(f)
        rows = (snap.get("filters") or {}).get("wallstreetbets") or []
        if not rows:
            continue
        dates.append(name[:-5])
        maps[name[:-5]] = {r["ticker"]: r["mentions"] for r in rows}
    return dates, maps


def scan(dates, maps, universe, floor, mult, look, gap):
    """precompute_events.compute, with the four constants as arguments.

    Deliberately a copy rather than an import: that module's constants are module-level
    by design — they are what the shipped artefact is built from — and making them
    parameters there would let a caller change what the pipeline produces.
    """
    events = []
    for tk in sorted(universe):
        last = -99
        for i in range(look, len(dates)):
            m = maps[dates[i]].get(tk, 0)
            med = st.median([maps[p].get(tk, 0) for p in dates[i - look:i]])
            if m >= floor and m >= mult * max(med, 1) and (i - last) > gap:
                events.append({"tk": tk, "d": dates[i], "m": m, "med": med})
                last = i
    events.sort(key=lambda x: x["d"], reverse=True)
    return events


def main():
    out_path = next((a for a in sys.argv[1:] if not a.startswith("-")), None)
    if not out_path:
        sys.exit(__doc__.strip().splitlines()[-3].strip())
    n = int(sys.argv[sys.argv.index("--n") + 1]) if "--n" in sys.argv else 40

    dates, maps = load_archive()
    universe = set()
    for d in dates:
        universe |= set(maps[d])
    with open(os.path.join(ROOT, "data", "prices.json")) as f:
        prices = json.load(f)
    with open(os.path.join(ROOT, "data", "days.json")) as f:
        days = json.load(f)

    random.seed(11)            # seeded: a rerun must compare the same settings
    cfgs, seen = list(CORNERS[:n]), set(CORNERS[:n])
    while len(cfgs) < n:
        c = (random.choice(FLOORS), random.choice(MULTS),
             random.randint(*LOOKS), random.randint(*GAPS))
        if c not in seen:
            seen.add(c)
            cfgs.append(c)

    out = {}
    for c in cfgs:
        events = scan(dates, maps, universe, *c)
        h = next((x for x in F.compute(events, prices, days)["horizons"]
                  if x["sessions"] == 20), None)
        out["-".join(map(str, c))] = {
            "n": len(events),
            "keys": sorted(e["tk"] + "|" + e["d"] for e in events),
            "spike": h["spike"] if h else None,
            "placebo": h["placebo"] if h else None,
            "gap_median": h["gap_median"] if h else None,
        }

    with open(out_path, "w") as f:
        json.dump(out, f)
    sizes = [v["n"] for v in out.values()]
    print(f"{len(out)} definitions · {min(sizes)}..{max(sizes)} events -> {out_path}")


if __name__ == "__main__":
    main()
